firebase-firestore/src/main/java/com/google/firebase/firestore/encoding/MapEncoder.java

```diff
@@ -0,0 +1,32 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.encoding;
+
+import java.util.Map;
+
+/**
+ * Provides an interface of encoding an custom object into a nested map of primitive types. This
```

Comment:
Describe in more detail what is expected of the classes implementing this interface. For example, who is going to use it and what restrictions it has over the values. For example, you mentioned that `encode` can produce maps of maps, but up to a maximum level. Are there any other restrictions?

Reply:
Originally I thought I should explain the behavior details at where I concrete implement this interface, like `MapEncoderKtxImp`, say maybe in the future, if we have another Java implementation, it may have different max depth rather than 500? But based on your other comments (to remove the todo), I think the goal is that we are most likely not going to have another subclass of this interface, so I have included more details at both here and `MapEncoderKtxImp`.